cpmpy/expressions/core.py

```diff
@@ -531,9 +534,69 @@ def value(self):
         elif self.name == "and": return all(arg_vals)
         elif self.name == "or" : return any(arg_vals)
         elif self.name == "->": return (not arg_vals[0]) or arg_vals[1]
+        elif self.name == "not": return not arg_vals[0]
 
         return None # default
 
+    def get_bounds(self):
+        if self.is_bool():
+            return 0, 1 #boolean
+        elif self.name == 'mul':
+            b1,b2 = get_bounds(self.args[0])
+            b3,b4 = get_bounds(self.args[1])
+            bounds = [b1 * b3, b1 * b4, b2 * b3, b2 * b4]
+            return min(bounds), max(bounds)
+        elif self.name == 'sum':
+            return sum([get_bounds(x)[0] for x in self.args]), sum([get_bounds(x)[1] for x in self.args])
+        elif self.name == 'wsum':
+            return sum(self.args[0] * np.array([get_bounds(x)[0] for x in self.args[1]])), \
+                   sum(self.args[0] * np.array([get_bounds(x)[1] for x in self.args[1]]))
+        elif self.name == 'sub':
+            return tuple(np.subtract(get_bounds(self.args[0]),get_bounds(self.args[1])))
+        elif self.name == 'div':
+            b1, b2 = get_bounds(self.args[0])
+            b3, b4 = get_bounds(self.args[1])
+            if b3 <= 0 <= b4:
+                raise ZeroDivisionError("division by domain containing 0 is not supported")
+            bounds = [b1 // b3, b1 // b4, b2 // b3, b2 // b4]
+            return min(bounds), max(bounds)
+        elif self.name == 'mod':
```

Comment:
This one is really tricky... Are we sure this covers all cases?

Reply:
pretty sure, also added tests now! wsum was wrong though